File: leetcode/unique-number-of-occurrences.java
// https://leetcode.com/problems/unique-number-of-occurrences/description/

import java.util.*;

class Solution {
    public boolean uniqueOccurrences(int[] arr) {
        HashSet<Integer> seen = new HashSet<Integer>();

        Arrays.sort(arr);

        int curr = arr[0];
        int occ = 1;
        for(int i = 1; i<arr.length; i++){
            if(arr[i] != curr){
                if(!seen.add(occ)) return false;
                occ = 0;
            }
            occ++;
            curr = arr[i];
        }

        return seen.add(occ);
    }
}

File: leetcode/unique_number_of_occurrences.py
# https://leetcode.com/problems/unique-number-of-occurrences


class Solution:
    def uniqueOccurrences(self, arr: list[int]) -> bool:
        arr.sort()

        curr_count: int = 1
        occurrences: set[int] = set()
        last: str = arr[0]

        for i in range(1, len(arr)):
            if arr[i] == last:
                curr_count += 1
            else:
                if curr_count in occurrences:
                    return False
                occurrences.add(curr_count)
                curr_count = 1
                last = arr[i]

        return curr_count not in occurrences
